contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,276 @@
+
+```go
+package api
+
+func WithAuditEvent(parent Context, e *audit.Event) Context
+func AuditEventFrom(ctx Context) (*audit.Event, bool)
+```
+
+Depending on the audit policy, different layers of the apiserver (e.g. http handler, storage) will fill the `audit.Event` struct. Certain fields might stay empty or `nil` if the policy does not require that field. E.g. in the case only http headers are supposed to be audit logged, no `OldObject` or `NewObject` is to be retrieved on the storage layer.
+
+The audit policy is a partial mapping from
+
+- kind,
+- namespace,
+- method,
+- and user
+
+to a policy action. An policy action defines the level of audit logging to be performed. The audit level can be
+
+- `HttpHeaders`,
+- `RequestObject`,
+- `StorageObject`.
+
+In addition the policy action can contain a number of output backend dependent key/values e.g. to define JSON object paths which should be logged or excluded from logging. For portable policy actions, a number of key/values are standardised and ought to be supported by output backends.
+
+When the http request is processed, the request handler will close the audit event by filling in the http response. Then it will pass the event to the configured policy backend.
+
+**Note:** for service creation and deletion there is special REST code in the apiserver which takes care of service/node port (de)allocation and removal of endpoints on service deletion. Hence, these operations are not visible on the API layer and cannot be audit logged therefore. **No other resources** (with the exception of componentstatus which is not of interest here) **implement this kind of custom CRUD operations.**
+
+### Events
+
+```go
+package audit
+
+// LogLevel defines the amount of information logged for auditing
+type LogLevel int
+
+
+type Event struct {
+  Level LogLevel
+
+  // http header level (HeaderLogLevel and higher)
+  ID string
+  Timestamp time.Timestamp
+  IP string
+  Method string
+  URI string
+  User string
+  AsUser string
+  Namespace, Name string
+  GroupKind unversioned.GroupKind
+  Response int
+
+  // CRUD level (RequestLogLevel and higher)
+  RequestObject runtime.Unstructured // before admission
+
+  // Storage level (StorageLogLevel and higher)
+  OldObject runtime.Object
+  NewObject runtime.Object
+  Patch     []byte
+}
+```
+
+### Concurrency/Transaction Model for Events
+
+TODO: define (and maybe adapt interface) how synchronization of the writable audit event of a context takes place.
+
+### Policy
+
+```go
+package audit
+
+const (
+  // DontLogLevel means to not log at all
+  DontLogLevel LogLevel = iota
+
+  // HeaderLogLevel means to log only information from the http headers, not from the body
+  HeaderLogLevel
+
+  // RequestLogLevel adds the user request in the body
+  RequestLogLevel
+
+  // StorageLogLevel adds the old and the new object that is received from and sent to the storage layer
+  StorageLogLevel
+)
+
+type Action interface {
```

Comment:
nit: Can this be a struct instead of an interface? I think it simplifies the model if it's assumed to be static (and lets us define more structured fields)